test/TestAppUtils/VisualTreeDumper.cs

```diff
@@ -0,0 +1,184 @@
+
+        interface IVisitor
+        {
+            bool ShouldVisitNode(DependencyObject node);
+            void BeginVisitNode(DependencyObject node);
+            void EndVisitNode(DependencyObject node);
+
+            bool ShouldVisitPropertiesForNode(DependencyObject node);
+            bool ShouldVisitProperty(PropertyInfo propertyInfo);
+            void VisitProperty(String propertyName, Object value);
+        }
+
+        class Visitor : IVisitor
+        {
+            private StringBuilder _sb;
+            private int _indent;
+            private IFilter _filter;
+            private IPropertyValueTranslator _translator;
+            public Visitor(IFilter filter, IPropertyValueTranslator translator)
+            {
+                _sb = new StringBuilder();
+                _sb.AppendLine("<?xml version=\"1.0\" encoding=\"utf-8\" ?>");
+                _indent = 0;
+                _filter = filter;
+                _translator = translator;
+            }
+            public void EndVisitNode(DependencyObject obj)
+            {
+                _indent--;
+                AddPadding(_indent);
+                _sb.AppendLine("</Element>");
+            }
+
+            public void BeginVisitNode(DependencyObject obj)
+            {
+                AddPadding(_indent);
+                _sb.AppendLine(String.Format("<Element Type=\"{0}\">", obj.GetType().FullName));
+                _indent++;
+            }
+
+            public override String ToString()
+            {
+                return _sb.ToString();
+            }
+
+            public bool ShouldVisitNode(DependencyObject node)
+            {
+                return node != null && _filter.ShouldLogElement(node.GetType().FullName);
+            }
+
+            public bool ShouldVisitPropertiesForNode(DependencyObject node)
+            {
+                return (node as UIElement) != null && _filter.ShouldLogElement(node.GetType().FullName);
+            }
+
+            public bool ShouldVisitProperty(PropertyInfo propertyInfo)
+            {
+                return _filter.ShouldLogProperty(propertyInfo.Name);
+            }
+            public void VisitProperty(string propertyName, object value)
+            {
+                var v = _translator.PropertyValueToString(propertyName, value);
+                if (!_filter.IsKnownProperty(propertyName, v))
+                {
+                    AddPadding(_indent + 1);
+                    _sb.AppendLine(String.Format("<Property Name=\"{0}\" Value=\"{1}\" />", propertyName, v));
+                }
+            }
+
+            private void AddPadding(int numOfSpace)
+            {
+                _sb.Append("".PadRight(numOfSpace));
+            }
+        }
+
+        public static String DumpToXML(DependencyObject root, IPropertyValueTranslator translator, IFilter filter)
+        {
+
+            Visitor visitor = new Visitor(filter ?? new DefaultFilter(), translator ?? new DefaultPropertyValueTranslator());
```

Comment:
Are these filters/translators stateful? Can you get by with a static instance? Filter.Default, for example

Reply:
I guess you just create once for the whole document, so it's not a big deal. I just get so used to seeing allocations and thinking about how we can avoid them :)

Reply:
I tried, and it doesn't work. because filter is interface, and the error is "interfaces can't contain fields"

---
In reply to: [254930374](https://github.com/Microsoft/microsoft-ui-xaml/pull/126#discussion_r254930374) [](ancestors = 254930374)